src/ManipulationMsg.chpl

```diff
@@ -232,6 +233,219 @@ module ManipulationMsg {
+    where ((t2 == int || t2 == bool) &&
+           (t == int || t == real || t == bool || t == uint || t == uint(8) || t == bigint) &&
+           (d2.rank == 1)) {
+    param pn = Reflection.getRoutineName();
+    
+    // Need to ensure we have a boolean array of what gets deleted.
+    var del_bool = makeDistArray(eIn.shape[axis], bool);
+    
+    if t2 == int {
+
+      // Error handling
+      var invalid_index: atomic bool = false;
+      var bad_index: atomic int = -1;
+      
+      forall i in del.domain with (
+        var agg = newDstAggregator(bool)
+      ) {
+
+        // Index out of bounds checking
+        if del[i] < eIn.shape[axis] && del[i] >= 0 {
+          agg.copy(del_bool[del[i]], true);
+        } else if del[i] >= -eIn.shape[axis] {
+          // numpy allows negative indices in this way
+          agg.copy(del_bool[del[i] + eIn.shape[axis]], true);
+        } else {
+          invalid_index.write(true);
+          bad_index.write(del[i]);
+        }
+        
+      }
+
+      if invalid_index.read() {
+        var errorMsg = incompatibleArgumentsError(pn, 
+                          "index %i is out of bounds for axis %i with size %i".format(bad_index.read(), axis, eIn.shape[axis])); 
+        mLogger.error(getModuleName(),getRoutineName(),getLineNumber(),errorMsg);                               
+        throw new Error(errorMsg);
+      }
+
+    } else {
+      if (del_bool.size != del.size) { 
+        var errorMsg = incompatibleArgumentsError(pn, 
+                          "Boolean array argument obj to delete must be one dimensional and match the axis length of %i".format(del_bool.size)); 
+        mLogger.error(getModuleName(),getRoutineName(),getLineNumber(),errorMsg);                               
+        throw new Error(errorMsg);
+      }
+      del_bool = del[..];
+    }
+
+    var eOut = makeDistArray((...deleteShape(eIn.shape, del_bool, axis)), t);
+
+    // Now true means it won't be deleted.
+    var not_del_bool: [del_bool.domain] bool;
+    forall i in del_bool.domain do not_del_bool[i] = !del_bool[i];
+
+    // Running sum of what isn't deleted, this increments whenever we hit something that stays
+    var sum_array: [del_bool.domain] int = + scan not_del_bool;
+
+    // This takes indices from eOut and converts them to indices in eIn
+    var map_array: [0..<(eIn.shape[axis] - (+ reduce del_bool))] int;
+    forall i in sum_array.domain {
+      if not_del_bool[i] then map_array[sum_array[i] - 1] = i;
+    }
+
+    // Performs the copy into eOut
+    forall idx in eOut.domain with (
+      var agg = newSrcAggregator(t),
+      ref map_array
+    ) {
+      var inIdx = if eIn.shape.size == 1 then (idx,) else idx;
+      inIdx[axis] = map_array[inIdx[axis]];
+      agg.copy(eOut[idx], eIn[if eIn.shape.size == 1 then inIdx[0] else inIdx]);
+    }
+
+    return eOut;
+  }
+
+  // This deletes by copying the slices that aren't deleted.
+  @arkouda.registerCommand(name="delete2")
```

Comment:
I wouldn't be against merging both implementations, but can you make the commands and the server functions have more self-documenting names? To me, the differentiating aspects are one is doing bulk copy and the other is doing aggregated copy.